core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java

```diff
@@ -866,6 +866,10 @@ private static void matchFilter(SubQueryRemoveRule rule,
           LogicVisitor.find(RelOptUtil.Logic.TRUE, ImmutableList.of(c), e);
       final Set<CorrelationId>  variablesSet =
           RelOptUtil.getVariablesUsed(e.rel);
+      // Only consider the correlated variables which originated from this sub-query level
+      if (!filter.getVariablesSet().isEmpty()) {
```

Comment:
NIT: how about add a comment for this condition to mention about what we've discussed, I think it's valuable for others to understand it.